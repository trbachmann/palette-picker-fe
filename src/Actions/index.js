export const setProjects = (projects) => ({
  type: 'SET_PROJECTS',
  projects
});

export const addProjectPalettes = (projectPalettes, id) => ({
  type: 'ADD_PROJECT_PALETTES',
  id,
  projectPalettes
})

export const setError = (error) => ({
  type: 'SET_ERROR',
  error
});

export const toggleLoading = (bool) => ({
  type: 'TOGGLE_LOADING',
  isLoading: bool
});

export const setPaletteColors = (colors) => ({
  type: 'SET_PALETTE_COLORS',
  palette: colors
});

export const setCurrentExpandedProject = (id) => ({
  type: 'SET_CURRENT_EXPANDED_PROJECT',
  id
})
export const addProject = (name, id) => ({
  type: 'ADD_PROJECT',
  name,
  id
});

export const addPaletteToProject = (projectId, palette) => ({
  type: 'ADD_PALETTE',
  projectId,
  palette
});

export const deletePalette = (paletteId, projectId) => ({
  type: 'DELETE_PALETTE',
  paletteId,
  projectId
});

export const deleteProject = (projectId) => ({
  type: 'DELETE_PROJECT',
  projectId
});

export const expandMiniPalette = (colors) => ({
  type: 'EXPAND_MINI_PALETTE',
  colors
});

export const replacePalette = (palette) => ({
  type: 'REPLACE_PALETTE',
  palette
});

export const updateProjectName = (name) => ({
  type: 'UPDATE_PROJECT_NAME',
  name
});