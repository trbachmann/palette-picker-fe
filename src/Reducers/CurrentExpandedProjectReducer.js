export const currentExpandedProjectReducer = (state = '', action) => {
  switch (action.type) {
    case 'SET_CURRENT_EXPANDED_PROJECT':
      return action.id
    default: 
      return state
  };
};