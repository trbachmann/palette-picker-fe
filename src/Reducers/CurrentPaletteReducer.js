export const currentPaletteReducer = (state = [], action) => {
  switch (action.type) {
    case 'SET_PALETTE':
      return action.palette
    default:
      return state
  };
};