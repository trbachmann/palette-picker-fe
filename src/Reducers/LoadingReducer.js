export const loadingReducer = (state = false, action) => {
  switch (action.type) {
    case 'TOGGLE_LOADING':
      return action.isLoading
    default:
      return state
  };
};