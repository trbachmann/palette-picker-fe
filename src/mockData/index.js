export const mockProjects = [
  {
    "id": 1,
    "name": "Game Time",
    "created_at": "2019-03-23T18:13:25.614Z",
    "updated_at": "2019-03-23T18:13:25.614Z"
  },
  {
    "id": 2,
    "name": "Movie Tracker",
    "created_at": "2019-03-23T18:13:25.631Z",
    "updated_at": "2019-03-23T18:13:25.631Z"
  },
  {
    "id": 3,
    "name": "Trapper Keeper",
    "created_at": "2019-03-23T18:13:25.652Z",
    "updated_at": "2019-03-23T18:13:25.652Z"
  }
];

export const mockProjectsWithEmptyPalettes = [
  {
    "id": 1,
    "name": "Game Time",
    "created_at": "2019-03-23T18:13:25.614Z",
    "updated_at": "2019-03-23T18:13:25.614Z",
    palettes: []
  },
  {
    "id": 2,
    "name": "Movie Tracker",
    "created_at": "2019-03-23T18:13:25.631Z",
    "updated_at": "2019-03-23T18:13:25.631Z",
    palettes: []
  },
  {
    "id": 3,
    "name": "Trapper Keeper",
    "created_at": "2019-03-23T18:13:25.652Z",
    "updated_at": "2019-03-23T18:13:25.652Z",
    palettes: []
  }
];

export const mockProjectsWithFullPalettes = [
  {
    "id": 1,
    "name": "Game Time",
    "created_at": "2019-03-23T18:13:25.614Z",
    "updated_at": "2019-03-23T18:13:25.614Z",
    palettes: [
      {
          id: 1,
          name: "Not favorites",
          color1: "#FCB97D",
          color2: "#EDD892",
          color3: "#C6B89E",
          color4: "#B5B8A3",
          color5: "#AABA9E",
          project_id: 1,
          created_at: "2019-03-19T20:28:06.194Z",
          updated_at: "2019-03-19T20:28:06.194Z"
      },
      {
          id: 2,
          name: "Favorites",
          color1: "#A0DDFF",
          color2: "#758ECD",
          color3: "#C1CEFE",
          color4: "#7189FF",
          color5: "#624CAB",
          project_id: 1,
          created_at: "2019-03-19T20:28:06.193Z",
          updated_at: "2019-03-19T20:28:06.193Z"
      }
    ]
  },
  {
    "id": 2,
    "name": "Movie Tracker",
    "created_at": "2019-03-23T18:13:25.631Z",
    "updated_at": "2019-03-23T18:13:25.631Z",
    palettes: []
  },
  {
    "id": 3,
    "name": "Trapper Keeper",
    "created_at": "2019-03-23T18:13:25.652Z",
    "updated_at": "2019-03-23T18:13:25.652Z",
    palettes: []
  }
];

export const mockPalette = [
  { hex: "#FCB97D", isLocked: false },
  { hex: "#EDD892", isLocked: false },
  { hex: "#C6B89E", isLocked: false },
  { hex: "#B5B8A3", isLocked: false },
  { hex: "#AABA9E", isLocked: false }
];

export const mockProjectPalettes = [
  {
      id: 1,
      name: "Not favorites",
      color1: "#FCB97D",
      color2: "#EDD892",
      color3: "#C6B89E",
      color4: "#B5B8A3",
      color5: "#AABA9E",
      project_id: 1,
      created_at: "2019-03-19T20:28:06.194Z",
      updated_at: "2019-03-19T20:28:06.194Z"
  },
  {
      id: 2,
      name: "Favorites",
      color1: "#A0DDFF",
      color2: "#758ECD",
      color3: "#C1CEFE",
      color4: "#7189FF",
      color5: "#624CAB",
      project_id: 1,
      created_at: "2019-03-19T20:28:06.193Z",
      updated_at: "2019-03-19T20:28:06.193Z"
  }
];

export const mockProjectsAfterDelete = [
  {
    "id": 1,
    "name": "Game Time",
    "created_at": "2019-03-23T18:13:25.614Z",
    "updated_at": "2019-03-23T18:13:25.614Z",
    palettes: [
      {
        id: 2,
        name: "Favorites",
        color1: "#A0DDFF",
        color2: "#758ECD",
        color3: "#C1CEFE",
        color4: "#7189FF",
        color5: "#624CAB",
        project_id: 1,
        created_at: "2019-03-19T20:28:06.193Z",
        updated_at: "2019-03-19T20:28:06.193Z"
      }
    ]
  },
  {
    "id": 2,
    "name": "Movie Tracker",
    "created_at": "2019-03-23T18:13:25.631Z",
    "updated_at": "2019-03-23T18:13:25.631Z",
    palettes: []
  },
  {
    "id": 3,
    "name": "Trapper Keeper",
    "created_at": "2019-03-23T18:13:25.652Z",
    "updated_at": "2019-03-23T18:13:25.652Z",
    palettes: []
  }
];

export const mockProjectsAfterProjectDelete = [
  {
    "id": 1,
    "name": "Game Time",
    "created_at": "2019-03-23T18:13:25.614Z",
    "updated_at": "2019-03-23T18:13:25.614Z"
  },
  {
    "id": 3,
    "name": "Trapper Keeper",
    "created_at": "2019-03-23T18:13:25.652Z",
    "updated_at": "2019-03-23T18:13:25.652Z"
  }
];

export const mockPaletteBeforeUpdate = {
  name: 'Great Greens',
  color1: "#FCB97D",
  color2: "#EDD892",
  color3: "#C6B89E",
  color4: "#B5B8A3",
  color5: "#AABA9E",
  project_id: 1
};

export const mockPaletteWithUpdates = {
  name: 'Great Green Shades',
  color1: "#GGGGGG",
  color2: "#EDD892",
  color3: "#C6B89E",
  color4: "#G2G2G3",
  color5: "#AABA9E",
  project_id: 1
};

export const mockProjectsBeforePaletteUpdate = [
  {
    id: 1,
    name: "Game Maker",
    created_at: "2019-03-23T18:13:25.614Z",
    updated_at: "2019-03-23T18:13:25.614Z",
    palettes: [
      mockPaletteBeforeUpdate
    ]
  },
  {
    id: 2,
    name: "Keeper Time",
    created_at: "2019-03-23T18:13:27.614Z",
    updated_at: "2019-03-23T18:13:27.614Z",
    palettes: []
  }
];

export const mockProjectsAfterPaletteUpdate = [
  {
    id: 1,
    name: "Game Maker",
    created_at: "2019-03-23T18:13:25.614Z",
    updated_at: "2019-03-23T18:13:25.614Z",
    palettes: [
      mockPaletteWithUpdates
    ]
  },
  {
    id: 2,
    name: "Keeper Time",
    created_at: "2019-03-23T18:13:27.614Z",
    updated_at: "2019-03-23T18:13:27.614Z",
    palettes: []
  }
];

export const mockProjectsAfterNameUpdate = [
  {
    "id": 1,
    "name": "Game Time",
    "created_at": "2019-03-23T18:13:25.614Z",
    "updated_at": "2019-03-23T18:13:25.614Z",
    palettes: []
  },
  {
    "id": 2,
    "name": "Movie Matcher",
    "created_at": "2019-03-23T18:13:25.631Z",
    "updated_at": "2019-03-23T18:13:25.631Z",
    palettes: []
  },
  {
    "id": 3,
    "name": "Trapper Keeper",
    "created_at": "2019-03-23T18:13:25.652Z",
    "updated_at": "2019-03-23T18:13:25.652Z",
    palettes: []
  }
];

export const mockMinPaletteToExpand = [
  "#FGFHES",
  "#E6D292",
  "#B6B89E",
  "#A5B8A3",
  "#DABA9E",
];

export const mockColorsAfterExpand = [
  { hex: "#FGFHES", isLocked: false },
  { hex: "#E6D292", isLocked: false },
  { hex: "#B6B89E", isLocked: false },
  { hex: "#A5B8A3", isLocked: false },
  { hex: "#DABA9E", isLocked: false }
];

export const mockPaletteAfterLock = [
  { hex: "#FCB97D", isLocked: false },
  { hex: "#EDD892", isLocked: false },
  { hex: "#C6B89E", isLocked: false },
  { hex: "#B5B8A3", isLocked: true },
  { hex: "#AABA9E", isLocked: false }
]