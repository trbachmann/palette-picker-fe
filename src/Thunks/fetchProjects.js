import { toggleLoading, setError, setProjects } from '../Actions';
import { fetchData } from '../Utils/fetch';

export const fetchProjects = () => {
  return async (dispatch) => {
    try {
      dispatch(toggleLoading(true));
      const projects = await fetchData('/projects');
      dispatch(toggleLoading(false));
      dispatch(setProjects(projects));
    } catch (error) {
      dispatch(setError(error));
    }
  }
};