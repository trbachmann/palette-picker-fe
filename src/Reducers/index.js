import React from 'react';
import { combineReducers } from 'redux';
import { currentPaletteReducer } from './CurrentPaletteReducer';
import { errorReducer } from './ErrorReducer';
import { loadingReducer } from './LoadingReducer';
import { projectsReducer } from './ProjectsReducer';
import { currentExpandedProjectReducer } from './CurrentExpandedProjectReducer';

export const rootReducer = combineReducers({
  currentPalette: currentPaletteReducer,
  error: errorReducer,
  isLoading: loadingReducer,
  projects: projectsReducer,
  currentExpandedProject: currentExpandedProjectReducer
});