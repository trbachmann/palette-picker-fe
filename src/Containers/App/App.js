import React, { Component } from 'react';
import PaletteArea from '../PaletteArea/PaletteArea';
import Sidebar from '../Sidebar/Sidebar';
import Error from '../../Components/Error/Error';
import Loading from '../../Components/Loading/Loading';
import PaletteForm from '../PaletteForm/PaletteForm';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { fetchProjects } from '../../Thunks/fetchProjects';
import { setPaletteColors } from '../../Actions'

export class App extends Component {
  componentDidMount() {
    this.props.fetchProjects();
    this.generatePalette();
  };

  generatePalette() {
    let palette = [];
    for(let i = 0; i < 5; i++) {
      let colorCode = this.generateColor()
      let color = { hex: colorCode, isLocked: false }
      palette.push(color)
    }
    this.props.setPaletteColors(palette)
  }

  generateColor() {
    let color = '#'
    for(let i=0; i<3; i++) {
      let hexPair = Math.floor(Math.random()*256).toString(16)
      if(hexPair.length < 2){
        hexPair = 0 + hexPair
      }
      color = color + hexPair
    }
    return color
  }

  render() {
    return (
      <div className="App">
        <h1>Colorations</h1>
        <button>Generate A Palette</button>
        <PaletteArea/>
        <PaletteForm/>
        <Sidebar/>
        <Error/>
        <Loading/>
      </div>
    );
  }
}

export const mapStateToProps = state => ({
  error: state.error,
  isLoading: state.isLoading
})

export const mapDispatchToProps = dispatch => ({
  fetchProjects: () => dispatch(fetchProjects()),
  setPaletteColors: (palette) => dispatch(setPaletteColors(palette))
})

export default connect(mapStateToProps, mapDispatchToProps)(App);
