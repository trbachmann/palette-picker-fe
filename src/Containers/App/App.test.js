import React from 'react';
import ReactDOM from 'react-dom';
import { App, mapStateToProps, mapDispatchToProps } from './App';
import { shallow } from 'enzyme';
import { Provider } from 'react-redux';
import { createStore } from 'redux';
import { rootReducer } from '../../Reducers';
import { fetchProjects } from '../../Thunks/fetchProjects';
import * as mockData from '../../mockData';
import { setPaletteColors } from '../../Actions';

jest.mock('../../Thunks/fetchProjects');

describe('App', () => {
  const mockProps = {
    error: '',
    isLoading: false,
    fetchProjects: jest.fn(),
    setPaletteColors: jest.fn(),
    currentPalette: mockData.mockPalette
  }

  it('renders without crashing', () => {
    const div = document.createElement('div');
    const store = createStore(rootReducer);
    ReactDOM.render(<Provider store={store}><App {...mockProps}/></Provider>, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  describe('App component', () => {
    let wrapper;
    
    beforeEach(() => {
      wrapper = shallow(<App {...mockProps}/>);
    });

    it('should match the snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('should call fetchProjects on componentDidMount', () => {
      wrapper.instance().componentDidMount();
      expect(mockProps.fetchProjects).toHaveBeenCalled();
    });
  });

  describe('mapStateToProps', () => {
    it('should return an object with an error and isLoading', () => {
      const mockState = {
        error: '',
        isLoading: false,
        projects: []
      }
      const expected = {
        error: '',
        isLoading: false
      }
      const mappedProps = mapStateToProps(mockState);
      expect(mappedProps).toEqual(expected);
    });
  });

  describe('mapDispatchToProps', () => {
    it('should call dispatch with fetchProjects', () => {
      const mockDispatch = jest.fn();
      const thunkToDispatch = fetchProjects();
      const mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.fetchProjects();
      expect(mockDispatch).toHaveBeenCalledWith(thunkToDispatch);
    });
  });
});