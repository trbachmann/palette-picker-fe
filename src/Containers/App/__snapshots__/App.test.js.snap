// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App App component should match the snapshot 1`] = `
<div
  className="App"
>
  <h1>
    Palette Picker
  </h1>
  <button>
    Generate A Palette
  </button>
  <PaletteArea />
  <PaletteForm />
  <Sidebar />
  <Error />
  <Loading />
</div>
`;
